# kcaoplot
R package for interstingğŸ­

>### install
```
library(devtools)
install_github("caokai001/kcaoplot")
library(kcaoplot)
```
---
# Part1 : try R package
> å°è¯•ä½¿ç”¨RåŒ…ï¼Œå†™å°å‡½æ•°ğŸ»

### example 1

test_ggplot(6)

![2019å¹´11æœˆ23æ—¥](https://upload-images.jianshu.io/upload_images/9589088-c0ea6db681989e36.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
- test_plot(4)
- hello()

>### contact: 
>[ç®€ä¹¦](https://www.jianshu.com/p/b5bc50d4c1d0)

---




# Part2: mutation analysis

> ç›®çš„ï¼š

â€‹	ä»ç™Œç—‡ç»„ç»‡ä¸­snp.vcf, å¯ä»¥çŒœæƒ³snp åˆ†å¸ƒä¹Ÿè®¸ä¸æ˜¯éšæœºçš„ï¼Œäºæ˜¯å¯ä»¥å°è¯•ç ”ç©¶æ˜¯å¦åˆ°æŸä¸ªmotifé™„è¿‘å¯Œé›†ã€‚

### example 2





> input file:

- â€‹          mynew.txt            :    Fimo.tsv file;use FIMO to scan hg19.fa
- NHEK_merge_ctcf.bed :    CTCF peak æ–‡ä»¶æˆ–è€…Dnase peak æ–‡ä»¶
- hg19_uv_snp_pos.bed ï¼š snp æ–‡ä»¶



```R
# UV
ctcf.motif=Fimo2GRanges("mynew.txt","GRanges")
NHEK.peak=bed2GRanges("./ctcf_peak/NHEK_merge_ctcf.bed",header=FALSE)
# snptogr
somatic_COLO829=bed2GRanges("hg19_uv_snp_pos.bed",header=FALSE)
names(mcols(somatic_COLO829)) <-c("Ref","Alt")
somatic_COLO829$id=c(1:length(somatic_COLO829))
# filter_fimo
motif.ovl=filter_fimo(NHEK.peak,ctcf.motif)
# æ‹“å®½åˆ°1000bp
filter_fimo_to_1000bp_gr(motif.ovl)
# snp è½åˆ°åŒºé—´ä½ç½®
mut.table<-mut_pos_tb(site,somatic_COLO829)
# ç”»å›¾1000bp
plot_mutation_loci(mut.table)
# ç”»å›¾flank 5bp
plot_flank_5bp(motif.ovl,somatic_COLO829)

```

![](https://upload-images.jianshu.io/upload_images/9589088-d4adfa1ca0722af2.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

### example 3

```R
# è¯»å…¥ctcf.motif
ctcf.motif=Fimo2GRanges("mynew.txt","GRanges")
# å¼€æ”¾åŒºåŸŸåˆå¹¶
dnase.peak<-bed2GRanges("E094-DNase.all.peaks.v2.bed")
dnase.fdr<-bed2GRanges("E094-DNase.fdr0.01.peaks.v2.bed")
dnase.mac<-bed2GRanges("E094-DNase.macs2.narrowPeak")
dnase=c(dnase.peak,dnase.fdr,dnase.mac)
<<<<<<< HEAD
=======
library(GenomicRanges)
>>>>>>> d47606c6f9c4fcbfee5d37b391c8c5fe3d5fe782
dnase=reduce(dnase)
# è¯»å…¥snpä¿¡æ¯
somatic_gastric<-bed2GRanges("41467_2018_3828_MOESM6_ESM.txt",header=TRUE)
names(mcols(somatic_gastric))[1:2]<-c("Ref","Alt")
somatic_gastric$id=c(1:length(somatic_gastric))
# è¿‡æ»¤FIMO
motif.ovl<-filter_fimo(dnase,ctcf.motif)
# æ‹“å±•åˆ°1kb
site=filter_fimo_to_1000bp_gr(motif.ovl)
# ä¸çªå˜ä½ç½®å–äº¤é›†
mut.table<-mut_pos_tb(site,somatic_gastric)
# ç”»1kp åŒºé—´çªå˜åˆ†å¸ƒå›¾
plot_mutation_loci(mut.table)
# ç”»flank_5bp åŒºé—´åˆ†å¸ƒå›¾
plot_flank_5bp(motif.ovl,somatic_gastric)
```

<img src="https://upload-images.jianshu.io/upload_images/9589088-45f2283c607b2157.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" style="zoom: 67%;" />


